### PR TITLE
Initial Crystal support

### PR DESCRIPTION
**Pull request by mamantoha**

### CODE REVIEW

**Comment by schani** [comment]
Wow, this looks amazingly complete!

**Comment by dvdsgl** [comment]
Please add `crystal` test fixture to `quicktype/.buildkite/pipeline.yml`.

**Comment by mamantoha** [comment]
@dvdsgl not sure about this file.

Will it be enough?

```yaml
  - command:
      "FIXTURE=crystal,scheme-crystal,csharp,schema-csharp,ruby,schema-ruby,golang,schema-golang .buildkite/build-pr.sh"
    label: "crystal csharp ruby golang"
```

**Comment by dvdsgl** [comment]
That's right.

**Comment by dvdsgl** [comment]
Great work! Looks almost ready to merge.

Your docs render oddly (see `[object Object]`):

```crystal
# Example code that deserializes and serializes the model.
# class [object Object]
#   include JSON::Serializable
#
#   @[JSON::Field(key: "answer")]
#   property answer : Int32
# end
#
# [object Object].from_json(%({"answer": 42}))
```

File: src/quicktype-core/language/Crystal.ts [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+        this.emitLine("alias ", name, " = ", this.crystalType(t));
+    }
+
+    protected emitLeadingComments(): void {
+        if (this.leadingComments !== undefined) {
+            this.emitCommentLines(this.leadingComments);
+            return;
+        }
+
+        const topLevelName = defined(iterableFirst(this.topLevels.keys()));
```
Comment (dvdsgl) [inline src/quicktype-core/language/Crystal.ts]: Take a look at `nameForNamedType` and `sourcelikeToString`:

```ruby
        const topLevel = defined(mapFirst(this.topLevels));
        const name = this.sourcelikeToString(this.nameForNamedType(topLevel));
        this.emitMultiline(
            `# Example code that deserializes and serializes the model.
# class ${name}
#   include JSON::Serializable
#
#   @[JSON::Field(key: "answer")]
#   property answer : Int32
# end
#
# ${this.sourcelikeToString(name)}.from_json(%({"answer": 42}))
`
        );
```

**Comment by dvdsgl** [comment]
Can you please run the test suite locally with:

```shell
$ FIXTURE=crystal,schema-crystal npm test
```

CI had an error:

```
Expected failure on input /app/test/inputs/schema/enum.1.fail.enum.json
--
  | {
  | "result": "{\"gve\":\"neutral\",\"lvc\":\"goody-two-shoes\"}\n",
  | "cwd": "/app/test/runs/schema-crystal-762e21"
  | }
  | Fixture threw an exception
  | {
  | "error": {
  | "result": "{\"gve\":\"neutral\",\"lvc\":\"goody-two-shoes\"}\n",
  | "cwd": "/app/test/runs/schema-crystal-762e21"
  | },
  | "sample": {
  | "path": "test/inputs/schema/enum.schema",
  | "additionalRendererOptions": {},
  | "saveOutput": true
  | },
  | "cwd": "/app/test/runs/schema-crystal-762e21"
  | }
```

File: src/quicktype-core/language/Crystal.ts [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+    protected makeRenderer(renderContext: RenderContext): CrystalRenderer {
+        return new CrystalRenderer(this, renderContext);
+    }
+
+    constructor() {
+        super("crystal", ["crystal", "cr", "crystallang"], "cr");
+    }
+
+    protected get defaultIndentation(): string {
+        return "  ";
```
Comment (schani) [inline src/quicktype-core/language/Crystal.ts]: Only two spaces?  Is that the convention in Crystal?

File: src/quicktype-core/language/Crystal.ts [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+                const [hasNull] = removeNullFromUnion(unionType);
+
+                const name = this.nameForNamedType(unionType);
+
+                return hasNull !== null ? ([name, "?"] as Sourcelike) : name;
+            }
+        );
+    };
+
+    private breakCycle = (t: Type, withIssues: boolean): any => {
```
Comment (schani) [inline src/quicktype-core/language/Crystal.ts]: This shouldn't be `any` but `Sourcelike`.

File: src/quicktype-core/language/Crystal.ts [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        const crystalType = this.crystalType(t, withIssues);
+        return crystalType;
+    };
+
+    private emitRenameAttribute(propName: Name, jsonName: string) {
```
Comment (schani) [inline src/quicktype-core/language/Crystal.ts]: Please type all methods that don't return values with `: void`.

File: src/quicktype-core/language/Crystal.ts [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+                this.emitDescription(this.descriptionForClassProperty(c, jsonName));
+                this.emitRenameAttribute(name, jsonName);
+                this.emitLine("property ", name, " : ", this.crystalType(prop.type, true));
+            });
+
+        this.emitBlock(["class ", className], structBody);
+    }
+
+    protected emitBlock(line: Sourcelike, f: () => void): void {
+        this.emitLine(line, "");
```
Comment (schani) [inline src/quicktype-core/language/Crystal.ts]: What's the `""` for here?

File: src/quicktype-core/language/Crystal.ts (hunk 2 above)
Comment (mamantoha) [reply src/quicktype-core/language/Crystal.ts]: Yes https://crystal-lang.org/docs/conventions/coding_style.html

**Comment by schani** [comment]
@mamantoha What's the problem with those skipped schema tests?

**Comment by schani** [comment]
Btw, don't worry about the failed `diff` test - it's not a failure, it's mainly me being too lazy to make CI not fail when new tests or languages are added.

**Comment by mamantoha** [comment]
@schani I left comments about skipped schema tests.

**Comment by schani** [comment]
@mamantoha Wonderful, thank you!  I'll merge it as soon as the tests pass again :-)

**Comment by schani** [comment]
Fantastic work, thank you so much!!!

**Comment by dvdsgl** [comment]
Bravo 👏

**Pull request merged**